File: examples/verify-receipt.mjs
#!/usr/bin/env node
// Offline L1 verification of a single VCC receipt, using the standalone
// TypeScript verifier in verifiers/typescript (build it once with
// `npm ci && npm run build` in that directory).
//
// Usage:
//   node examples/verify-receipt.mjs <receipt.json> <keyset.json>
//
// <receipt.json> may be a `{ statement, envelope }` receipt (as in examples/)
// or a bare DSSE envelope. <keyset.json> may be a `{ keyset: ... }` document
// (as in vectors/test-key.json or /.well-known/vcc-issuer.json) or a bare keyset.
//
// Prints the verification AXES separately — there is deliberately no single
// "verified" boolean anywhere in VCC. Exit 0 iff cryptographicValidity.
// No network calls are made.

import { readFileSync } from "node:fs";
import { verifyVccEnvelope } from "../verifiers/typescript/dist/src/index.js";

const [receiptPath, keysetPath] = process.argv.slice(2);
if (!receiptPath || !keysetPath) {
  console.error("usage: node examples/verify-receipt.mjs <receipt.json> <keyset.json>");
  process.exit(2);
}

const receiptDoc = JSON.parse(readFileSync(receiptPath, "utf8"));
const envelope = receiptDoc.envelope ?? receiptDoc;
const keysetDoc = JSON.parse(readFileSync(keysetPath, "utf8"));
const keyset = keysetDoc.keyset ?? keysetDoc;

const res = await verifyVccEnvelope(envelope, keyset);

console.log("axes:");
console.log(`  signatureValid            : ${res.signatureValid}`);
console.log(`  statementIntact           : ${res.statementIntact}`);
console.log(`  issuerIdentityBound       : ${res.issuerIdentityBound}`);
console.log(`  keyValidAtIssuance        : ${res.keyValidAtIssuance}`);
console.log(`  issuerKeyStatus           : ${res.issuerKeyStatus}`);
console.log(`  certificateStatus         : ${res.certificateStatus}`);
console.log(`  cryptographicValidity     : ${res.cryptographicValidity}`);
console.log(`  trustedAtVerificationTime : ${res.trustedAtVerificationTime}`);
console.log(`checks: ${JSON.stringify(res.checks)}`);
if (res.errors.length > 0) {
  console.log(`errors: ${JSON.stringify(res.errors)}`);
}
process.exit(res.cryptographicValidity ? 0 : 1);

File: examples/verify_receipt.py
#!/usr/bin/env python3
"""Offline L1 verification of a single VCC receipt, using the independent
Python verifier in verifiers/python (install its single dependency once:
`python3 -m venv verifiers/python/.venv &&
verifiers/python/.venv/bin/pip install -r verifiers/python/requirements.txt`).

Usage:
    python3 examples/verify_receipt.py <receipt.json> <keyset.json>

<receipt.json> may be a `{ statement, envelope }` receipt (as in examples/)
or a bare DSSE envelope. <keyset.json> may be a `{ keyset: ... }` document
(as in vectors/test-key.json or /.well-known/vcc-issuer.json) or a bare keyset.

Prints the verification AXES separately — there is deliberately no single
"verified" boolean anywhere in VCC. Exit 0 iff cryptographicValidity.
No network calls are made.
"""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "verifiers" / "python"))
from vcc_verifier import verify_vcc_envelope  # noqa: E402


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: python3 examples/verify_receipt.py <receipt.json> <keyset.json>", file=sys.stderr)
        return 2

    receipt_doc = json.loads(Path(sys.argv[1]).read_text(encoding="utf-8"))
    envelope = receipt_doc.get("envelope", receipt_doc)
    keyset_doc = json.loads(Path(sys.argv[2]).read_text(encoding="utf-8"))
    keyset = keyset_doc.get("keyset", keyset_doc)

    res = verify_vcc_envelope(envelope, keyset)

    print("axes:")
    print(f"  signatureValid            : {res.signatureValid}")
    print(f"  statementIntact           : {res.statementIntact}")
    print(f"  issuerIdentityBound       : {res.issuerIdentityBound}")
    print(f"  keyValidAtIssuance        : {res.keyValidAtIssuance}")
    print(f"  issuerKeyStatus           : {res.issuerKeyStatus}")
    print(f"  certificateStatus         : {res.certificateStatus}")
    print(f"  cryptographicValidity     : {res.cryptographicValidity}")
    print(f"  trustedAtVerificationTime : {res.trustedAtVerificationTime}")
    print(f"checks: {json.dumps(res.checks)}")
    if res.errors:
        print(f"errors: {json.dumps(res.errors)}")
    return 0 if res.cryptographicValidity else 1


if __name__ == "__main__":
    raise SystemExit(main())
